lib/fluent/plugin/output.rb

```diff
@@ -1222,29 +1222,29 @@ def update_retry_state(chunk_id, using_secondary, error = nil)
 
           # @retry exists
 
-          if error
-            if @retry.limit?
+          if @retry.limit?
+            if error
               records = @buffer.queued_records
               msg = "failed to flush the buffer, and hit limit for retries. dropping all chunks in the buffer queue."
               log.error msg, retry_times: @retry.steps, records: records, error: error
               log.error_backtrace error.backtrace
-            elsif using_secondary
-              msg = "failed to flush the buffer with secondary output."
-              log.warn msg, retry_time: @retry.steps, next_retry_seconds: @retry.next_time, chunk: chunk_id_hex, error: error
-              log.warn_backtrace error.backtrace
-            else
-              msg = "failed to flush the buffer."
-              log.warn msg, retry_time: @retry.steps, next_retry_seconds: @retry.next_time, chunk: chunk_id_hex, error: error
-              log.warn_backtrace error.backtrace
             end
-          end
-
-          if @retry.limit?
             @buffer.clear_queue!
             log.debug "buffer queue cleared"
             @retry = nil
           else
             @retry.step
+            if error
+              if using_secondary
+                msg = "failed to flush the buffer with secondary output."
+                log.warn msg, retry_time: @retry.steps, next_retry_seconds: @retry.next_time, chunk: chunk_id_hex, error: error
```

Comment:
Do we need to add `- 1` to `retry_time`?

Reply:
When `write` fails for the first time, the `retry_time` starts from 0.
Rather, `retry_time` was displayed as 0 even though it was the first retry.
I think `retry_time=0` should not be output twice.
(In [#2201](https://github.com/fluent/fluentd/issues/2201),  the logs was not enough, I am sorry.)
```
2018-12-06 04:12:29 +0000 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2018-12-06 04:12:30 +0000 chunk="57c52b4ef75540aa1bc46190486ae4b8" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"xxx.xxx.xxx.xxx\" port 24224"
  2018-12-06 04:12:29 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-1.2.6/lib/fluent/plugin_helper/socket.rb:59:in `initialize'
  2018-12-06 04:12:29 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-1.2.6/lib/fluent/plugin_helper/socket.rb:59:in `new'
...
2018-12-06 04:12:30 +0000 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2018-12-06 04:12:30 +0000 chunk="57c52b4ef75540aa1bc46190486ae4b8" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"xxx.xxx.xxx.xxx\" port 24224"
2018-12-06 04:12:31 +0000 [warn]: #0 failed to flush the buffer. retry_time=1 next_retry_seconds=2018-12-06 04:12:31 +0000 chunk="57c52b4ef75540aa1bc46190486ae4b8" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"xxx.xxx.xxx.xxx\" port 24224"
2018-12-06 04:12:33 +0000 [warn]: #0 failed to flush the buffer. retry_time=2 next_retry_seconds=2018-12-06 04:12:33 +0000 chunk="57c52b4ef75540aa1bc46190486ae4b8" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"xxx.xxx.xxx.xxx\" port 24224"
2018-12-06 04:12:37 +0000 [warn]: #0 failed to flush the buffer. retry_time=3 next_retry_seconds=2018-12-06 04:12:37 +0000 chunk="57c52b4ef75540aa1bc46190486ae4b8" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"xxx.xxx.xxx.xxx\" port 24224"
```

Reply:
> Do we need to add - 1 to retry_time?

No. `retry_time` starts from 0. Display `retry_time=0` when Fluentd calls `try_flush` first time.

In this version, logs like following:

```
2018-12-06 14:46:29 +0900 [info]: #0 fluentd worker is now running worker=0
2018-12-06 14:46:31 +0900 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2018-12-06 14:46:32 +0900 chunk="57
c5405353f5e55d61ab8bfaad92b44d" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-06 14:46:31 +0900 [warn]: #0 /home/kenji/ruby/td-family/fluentd/lib/fluent/plugin_helper/socket.rb:59:in `initialize'
  (snip)
2018-12-06 14:46:32 +0900 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2018-12-06 14:46:32 +0900 chunk="57c5405353f5e55d61ab8bfaad92b44d" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-06 14:46:32 +0900 [warn]: #0 suppressed same stacktrace
2018-12-06 14:46:33 +0900 [warn]: #0 failed to flush the buffer. retry_time=1 next_retry_seconds=2018-12-06 14:46:33 +0900 chunk="57c5405353f5e55d61ab8bfaad92b44d" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-06 14:46:33 +0900 [warn]: #0 suppressed same stacktrace
2018-12-06 14:46:35 +0900 [warn]: #0 failed to flush the buffer. retry_time=2 next_retry_seconds=2018-12-06 14:46:35 +0900 chunk="57c5405353f5e55d61ab8bfaad92b44d" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-06 14:46:35 +0900 [warn]: #0 suppressed same stacktrace

```

Reply:
Hmm... I see.
Looks like printing `retry_time=0` twice and same next_retry_seconds are weird.
`retry_time` should be start 1?

Reply:
Sorry. I've pasted wrong logs...

```
2018-12-10 10:43:28 +0900 [warn]: #0 failed to flush the buffer. retry_time=0 next_retry_seconds=2018-12-10 10:43:29 +0900 chunk="57
ca1174ccb6423c5465ec3e1b696a98" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
(snip)
2018-12-10 10:43:29 +0900 [warn]: #0 failed to flush the buffer. retry_time=1 next_retry_seconds=2018-12-10 10:43:30 +0900 chunk="57ca1174ccb6423c5465ec3e1b696a98" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-10 10:43:29 +0900 [warn]: #0 suppressed same stacktrace
2018-12-10 10:43:30 +0900 [warn]: #0 failed to flush the buffer. retry_time=2 next_retry_seconds=2018-12-10 10:43:32 +0900 chunk="57ca1174ccb6423c5465ec3e1b696a98" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-10 10:43:30 +0900 [warn]: #0 suppressed same stacktrace
2018-12-10 10:43:32 +0900 [warn]: #0 failed to flush the buffer. retry_time=3 next_retry_seconds=2018-12-10 10:43:36 +0900 chunk="57ca1174ccb6423c5465ec3e1b696a98" error_class=Errno::ECONNREFUSED error="Connection refused - connect(2) for \"::1\" port 24224"
  2018-12-10 10:43:32 +0900 [warn]: #0 suppressed same stacktrace
```